[CSApply/Distributed] Identify implicitly throwing calls while applying solution

xedin opened the pull request:
Calls to distributed actor methods could be pointing to a thunk for remote access,
such thunks are implicitly throwing. Their existence has to be identified early in order
for later steps of solution application to mark context as throwing e.g. implicitly
generated autoclosures for `async let` initializers rely on that to make closure as
throwing when necessary.

Resolves: rdar://83610106
<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Add accessor to check whether target represents `async let` init

xedin committed:
[Concurrency] Extract the check for async context to the namespace

It's useful not only inside of isolation checking but in the
type-checker as well.

xedin committed:
[Distributed] Add a way to determine if particular reference is to distributed thunk

Make it possible for type-checker to determine whether the given
reference in the given context represents a call to a remote distributed
actor.

xedin committed:
[CSApply] Mark call to distributed thunks as implicitly throwing

Use newly added `isDistributedThunk` check to determine whether
the given call is to a remote distributed actor and if so mark
it as implicitly throwing.

Resolves: rdar://83610106

xedin committed:
[Distributed] NFC: Add a test-case for different `async let` configurations

xedin commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -2696,6 +2688,53 @@ bool ActorIsolationChecker::mayExecuteConcurrentlyWith(
+  case ActorIsolationRestriction::ActorSelf: {
+    // An instance member of an actor can be referenced from an actor's
+    // designated initializer or deinitializer.
+    if (actor.isActorSelf() && !isInAsyncLetInitializer) {
```
I think this check is appropriate for both `ActorSelf` and `CrossActorSelf` because it means that reference is: ```swift
distributed actor {
   init(transport: AnyTranport) {
      _ = self.test()
   }

   distributed func test() -> Int {  42 }
}
```
The code in `TypeCheckConcurrency` isn't clear on that regard...

ktoso replied:
Let's double check with @kavon; this is "just" an actor function call on self from a synchronous init so it'll behave the same as in actors -- which Kavon is working on now

kavon replied:
These changes look fine to me. The special checking for actor inits are mainly happening in SIL.

xedin replied:
Thank you!

ktoso commented on lib/Sema/CSApply.cpp:
```diff
@@ -7581,6 +7581,15 @@ Expr *ExprRewriter::finishApply(ApplyExpr *apply, Type openedType,
+      auto *FD = cast<AbstractFunctionDecl>(callee.getDecl());
+      if (!FD->hasThrows())
+        apply->setImplicitlyThrows(true);
+    }
```
👍

ktoso commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -2696,6 +2688,53 @@ bool ActorIsolationChecker::mayExecuteConcurrentlyWith(
+  default:
+    return false;
+  }
+}
```
Looks good AFAICS... thank you!

ktoso reviewed (approved):
LGTM, thanks a lot @xedin! 💯

xedin commented:
@swift-ci please clean test

DougGregor commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -4282,3 +4326,27 @@ AnyFunctionType *swift::adjustFunctionTypeForConcurrency(
+
+  if (auto closure = dyn_cast<AbstractClosureExpr>(dc)) {
+    if (auto type = closure->getType()) {
+      if (auto fnType = type->getAs<AnyFunctionType>()) {
```
When we're inside the constraint solver, this will give the wrong answer because the closures we encounter won't have types yet. Rather, we will need to look up the closure's type within the constraint system to see if it's `async`.

xedin replied:
Good catch! I have overlooked this one, I guess we need a callback to get a type of a closure here...

xedin replied:
Fixed!

DougGregor commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -2696,6 +2688,58 @@ bool ActorIsolationChecker::mayExecuteConcurrentlyWith(
+  if (!isa<SelfApplyExpr>(context))
+    return false;
+
+  auto actor = getIsolatedActor(cast<SelfApplyExpr>(context)->getBase());
```
There's some Stateful Stuff :tm: going on in `getIsolatedActor` with closures, where it is looking at closure captures... but those haven't been computed yet if we're in the constraint solver. I suspect that's going to trip things up here.

xedin replied:
Opted to only checking whether actor is potentially isolated.

DougGregor reviewed (commented):
This is a great direction. I think we're going to need to pull more information into the constraint system to get actor-isolation checking working properly for closures when isolation is queried from inside the constraint solver like this; see my comments inline.

xedin committed:
[AST] Associated `isActorSelf` with `VarDecl`

xedin committed:
[TypeChecker] Implement distributed thunk identification in expr rewritter

This logic cannot live in `ActorIsolationChecker` because
all of the relevant information is only accessible through
constraint system while applying solutions.

xedin committed:
[TypeCheckConcurrency] Stop exposing `isAsynchronousContext` as no longer useful

`ConstraintSystem` has its own method to determine this.

DougGregor reviewed (approved):
This looks like it'll work, thank you!

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please test Windows platform

ktoso commented:
@swift-ci please test Windows platform

ktoso commented:
Awesome! And CI green everywhere as well now 🥳
